src/nvim/ui_compositor.c

```diff
@@ -0,0 +1,445 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
+// it. PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+
+// UI wrapper that sends requests to the UI thread.
+// Used by the built-in TUI and libnvim-based UIs.
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <limits.h>
+
+#include "nvim/lib/kvec.h"
+#include "nvim/log.h"
+#include "nvim/main.h"
+#include "nvim/ascii.h"
+#include "nvim/vim.h"
+#include "nvim/ui.h"
+#include "nvim/highlight.h"
+#include "nvim/memory.h"
+#include "nvim/ui_compositor.h"
+#include "nvim/ugrid.h"
+#include "nvim/screen.h"
+#include "nvim/syntax.h"
+#include "nvim/api/private/helpers.h"
+#include "nvim/os/os.h"
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "ui_compositor.c.generated.h"
+#endif
+
+#define UI(b) (((UICompositorData *)b)->ui)
+
+
+int composed_uis = 0;
+
+static UI *compositor = NULL;
+
+static ScreenGrid *curgrid;
+
+kvec_t(ScreenGrid *) layers = KV_INITIAL_VALUE;
+
+static size_t bufsize = 0;
+static schar_T *linebuf;
+static sattr_T *attrbuf;
+
+bool debug_recompose = false;
+
+#ifndef NDEBUG
+static int chk_width = 0, chk_height = 0;
+#endif
+
+static bool valid_screen = true;
+static bool msg_scroll_mode = false;
+static int msg_first_invalid = 0;
+
+void ui_comp_init(void)
+{
+  if (compositor != NULL) {
+    return;
+  }
+  compositor = xcalloc(1, sizeof(UI));
+
+  compositor->rgb = true;
+  compositor->grid_resize = ui_comp_grid_resize;
+  compositor->grid_clear = ui_comp_grid_clear;
+  compositor->grid_scroll = ui_comp_grid_scroll;
+  compositor->grid_cursor_goto = ui_comp_grid_cursor_goto;
+  compositor->raw_line = ui_comp_raw_line;
+  compositor->win_scroll_over_start = ui_comp_win_scroll_over_start;
+  compositor->win_scroll_over_reset = ui_comp_win_scroll_over_reset;
+  compositor->flush = ui_comp_flush;
+
+  // Be unoptionated: will be attached together with a "real" ui anyway
```

Comment:
should it be "unopinionated" or is it referring to options?

Reply:
unopinionated is better indeed :P